# **Project 1 - Sales Analysis - ETL**

## Objectives

- Extract data from provided CSV files
- Clean it 
- Apply feature engineering if necessary 
- Remove any unnecessary columns
- Save to a new CSV file


## Inputs

CSV files provided:

stores data-set.csv
sales data-set.csv
features data set.csv

Renamed files to a uniform naming convention:

Sales_Features_DataSet.csv
Sales_DataSet.csv
Sales_Stores_DataSet.csv

Note: original files are stored in Data/OriginalFiles


## Outputs

CSV file created from ETL etc stored in Data/CleanedDataSets:

Sales_Features_DataSet_Cleaned.csv



## Additional Comments

Had an error trying to use pip install - r requirements.txt 

So manually installed:

numpy
pandas
seaborn
plotly
seaborn
matplotlib

Will install any more required libraries as discovered...

Error:

(.venv) rogerwilliams@Rogers-iMac Project1-SalesAnalysis % pip install -r requirements.txt
Collecting numpy==1.26.1 (from -r requirements.txt (line 1))
  Using cached numpy-1.26.1-cp312-cp312-macosx_10_9_x86_64.whl.metadata (61 kB)
Collecting pandas==2.1.1 (from -r requirements.txt (line 2))
  Using cached pandas-2.1.1-cp312-cp312-macosx_10_9_x86_64.whl.metadata (18 kB)
Collecting matplotlib==3.8.0 (from -r requirements.txt (line 3))
  Using cached matplotlib-3.8.0-cp312-cp312-macosx_10_12_x86_64.whl.metadata (5.8 kB)
Collecting seaborn==0.13.2 (from -r requirements.txt (line 4))
  Using cached seaborn-0.13.2-py3-none-any.whl.metadata (5.4 kB)
Collecting ydata-profiling==4.12.0 (from -r requirements.txt (line 5))
  Using cached ydata_profiling-4.12.0-py2.py3-none-any.whl.metadata (20 kB)
Collecting plotly==5.17.0 (from -r requirements.txt (line 6))
  Using cached plotly-5.17.0-py2.py3-none-any.whl.metadata (7.0 kB)
Collecting ppscore==1.1.0 (from -r requirements.txt (line 7))
  Using cached ppscore-1.1.0.tar.gz (46 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [20 lines of output]
      Traceback (most recent call last):
        File "/Users/rogerwilliams/Projects/Python/CourseProjects/Project1-SalesAnalysis/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/Users/rogerwilliams/Projects/Python/CourseProjects/Project1-SalesAnalysis/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/rogerwilliams/Projects/Python/CourseProjects/Project1-SalesAnalysis/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/wv/yty584vs1kq76d7zd3rv5wpw0000gn/T/pip-build-env-4o10ho_h/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 333, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/wv/yty584vs1kq76d7zd3rv5wpw0000gn/T/pip-build-env-4o10ho_h/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 301, in _get_build_requires
          self.run_setup()
        File "/private/var/folders/wv/yty584vs1kq76d7zd3rv5wpw0000gn/T/pip-build-env-4o10ho_h/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 520, in run_setup
          super().run_setup(setup_script=setup_script)
        File "/private/var/folders/wv/yty584vs1kq76d7zd3rv5wpw0000gn/T/pip-build-env-4o10ho_h/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 317, in run_setup
          exec(code, locals())
        File "<string>", line 12, in <module>
      ModuleNotFoundError: No module named 'pkg_resources'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.

[notice] A new release of pip is available: 24.3.1 -> 26.1.2
[notice] To update, run: pip install --upgrade pip
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

Note: This error originates from a subprocess, and is likely not a problem with pip.


## Initalise Working Environment

In [21]:
#import libraries
import os
import numpy as np
import pandas as pd
import plotly.express as px
import seaborn as sns
import matplotlib.pyplot as plt


# Section 1 - Initalisation

## Intitialise all variables to be used in global stack 

In [22]:
#DataFrame vars for ETL
dfSales_Features_DataSet = None
dfSales_Features_DataSet_Work = None
dfCleaned = None
dfCheck = None

#missing values check vars
intMissing = 0
intMissingPercent = 0
intLessThanZero = 0
dfCheck = None

#duplicates check vars
intDuplicates = 0

#stores current directory
strCurrentDir = ""

#constant for project directory
CNST_STR_PROJECT_DIR = "Project1-SalesAnalysis"

#file path constants
CNST_STR_CLEANFILEPATH = "/Data/CleanedDataSets/Sales_Features_DataSet_Cleaned.csv"
CNST_STR_SALESFEATURESFILEPATH = "/Data/Sales_Features_DataSet.csv"

#other vars
objColumns = None
lstColumns = []


## Set Current Directory To Base Project Directory

In [23]:
#get project directory - default is jupyter notebook sub folder as that is where this file is located!
#so move back one to the project root path
# Source - https://stackoverflow.com/a/17726833
# Posted by chimpsarehungry
# Retrieved 2026-07-05, License - CC BY-SA 3.0

#get current folder
strCurrentDir = os.getcwd()

#is the last part of the path the project directory?
if not strCurrentDir.endswith(CNST_STR_PROJECT_DIR):
   #get current working directory and move back one to the project root path
   strCurrentDir =  os.path.normpath(os.getcwd() + os.sep + os.pardir)
   os.chdir(os.path.dirname(strCurrentDir))
   #change directory
   os.chdir(strCurrentDir)

#confirm current directory is project directory
print(f"Current Directory: \n {os.getcwd()}")

Current Directory: 
 /Users/rogerwilliams/Projects/Python/CourseProjects/Project1-SalesAnalysis


# Section 1 - Extraction

- Read csv files
- Get schema info -> column and row numbers
- Get first 5 records
- Get list of column datatypes
- Get detailed schema infomation

## Read csv Files Into Variable For Processing

In [24]:
#read csv file into DataFrame
dfSales_Features_DataSet = pd.read_csv(strCurrentDir + CNST_STR_SALESFEATURESFILEPATH)


## Get Schema Info - Sales_Features_DataSet

In [25]:
print(f"Schema Information \n Rows: {dfSales_Features_DataSet.shape[0]} Columns: {dfSales_Features_DataSet.shape[1]}")
dfSales_Features_DataSet.head()

Schema Information 
 Rows: 8190 Columns: 12


,Store,Date,Temperature,Fuel_Price,MarkDown1,MarkDown2,MarkDown3,MarkDown4,MarkDown5,CPI,Unemployment,IsHoliday
0,1,05/02/2010,42.31,2.572,NaN,NaN,NaN,NaN,NaN,211.096358,8.106,False
1,1,12/02/2010,38.51,2.548,NaN,NaN,NaN,NaN,NaN,211.242170,8.106,True
2,1,19/02/2010,39.93,2.514,NaN,NaN,NaN,NaN,NaN,211.289143,8.106,False
3,1,26/02/2010,46.63,2.561,NaN,NaN,NaN,NaN,NaN,211.319643,8.106,False
4,1,05/03/2010,46.50,2.625,NaN,NaN,NaN,NaN,NaN,211.350143,8.106,False


## Observations - Sales_Features_DataSet
8190 rows
12 columns

Columns:

Store, Date, Temperature, Fuel_Price, MarkDown1, MarkDown2, MarkDown3, MarkDown4, MarkDown5, 
CPI, Unemployment, IsHoliday

In [26]:
#get column data types
print (f"Data Types: \n{dfSales_Features_DataSet.dtypes}")

Data Types: 
Store             int64
Date                str
Temperature     float64
Fuel_Price      float64
MarkDown1       float64
MarkDown2       float64
MarkDown3       float64
MarkDown4       float64
MarkDown5       float64
CPI             float64
Unemployment    float64
IsHoliday          bool
dtype: object


## Observations - Sales_Features_DataSet

Date is a string - needs conversion to datetime
Other column types look ok

In [27]:
#get schema infomation
print(f"More Information:\n {dfSales_Features_DataSet.info()}")

<class 'pandas.DataFrame'>
RangeIndex: 8190 entries, 0 to 8189
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Store         8190 non-null   int64  
 1   Date          8190 non-null   str    
 2   Temperature   8190 non-null   float64
 3   Fuel_Price    8190 non-null   float64
 4   MarkDown1     4032 non-null   float64
 5   MarkDown2     2921 non-null   float64
 6   MarkDown3     3613 non-null   float64
 7   MarkDown4     3464 non-null   float64
 8   MarkDown5     4050 non-null   float64
 9   CPI           7605 non-null   float64
 10  Unemployment  7605 non-null   float64
 11  IsHoliday     8190 non-null   bool   
dtypes: bool(1), float64(9), int64(1), str(1)
memory usage: 712.0 KB
More Information:
 None


## Observations - Sales_Features_DataSet

No missing values
File size: 712 KB

---

# Section 2 - Transformation

- Get missing values percent
- Get duplicates (if any)


In [28]:
#get total missing values
intMissing = dfSales_Features_DataSet.isnull().sum()

#calculate percent missing
intMissingPercent = ( intMissing / len(dfSales_Features_DataSet) * 100)
dfCheck = pd.DataFrame( {"Missing Values": intMissing, "%" : intMissingPercent})

#show results
if dfCheck["Missing Values"].sum() !=0:
   print( dfCheck[dfCheck["Missing Values"] > 0])
else:
   print( "No Missing Values")     



              Missing Values          %
MarkDown1               4158  50.769231
MarkDown2               5269  64.334554
MarkDown3               4577  55.885226
MarkDown4               4726  57.704518
MarkDown5               4140  50.549451
CPI                      585   7.142857
Unemployment             585   7.142857


## Observations - Sales_Features_DataSet

Missing values for columns:
MarkDown1, MarkDown2, MarkDown3, MarkDown4, MarkDown5
CPI, Employment

CPI and Employment are not used in this analysis so will be dropped from the dataset later on

## Find Duplicates - Sales_Features_DataSet

In [29]:
#get number of duplicates
intDuplicates = dfSales_Features_DataSet.isnull().sum()
intDuplicates


Store              0
Date               0
Temperature        0
Fuel_Price         0
MarkDown1       4158
MarkDown2       5269
MarkDown3       4577
MarkDown4       4726
MarkDown5       4140
CPI              585
Unemployment     585
IsHoliday          0
dtype: int64

## Observations - Sales_Features_DataSet

Duplicates, but for numerical fields where could be expected
Plus there are a lot of missing values in these columns at the moment so to be expected

## Get Unique Values - Sales_Features_DataSet

In [30]:
#show unique values count
dfSales_Features_DataSet.nunique()

Store             45
Date             182
Temperature     4178
Fuel_Price      1011
MarkDown1       4023
MarkDown2       2715
MarkDown3       2885
MarkDown4       3405
MarkDown5       4045
CPI             2505
Unemployment     404
IsHoliday          2
dtype: int64

## Observations - Sales_Features_DataSet

Store has 45 unique values  
Date has 182 unique values  
Temperature has 4178 unique values
Fuel_Price has 1011 unique values
MarkDown1 has 4023 unique values
MarkDown2 has 2715 unique values
MarkDown3 has 2885 unique values
MarkDown4 has 3405 unique values
MarkDown5 has 4045 unique values
CPI has 2505 unique values
Unemployment has 404 unique values
IsHoliday has 2 unique values



## Get Categorical Value Distribution

In [31]:
#get categorical distribution  no needed as only one string value: Date
#and that will converted to datetime later on


In [32]:
## Create Copy Of DataFrame
dfSales_Features_DataSet_Work = dfSales_Features_DataSet.copy()

## Replace Missing Values With ZERO For MarkDown Columns

In [33]:
#replace missing values for markdown columns with zero Copilot code unchanged
#would have liked proper naming convention such as: lstMarkDownColumns
markdown_cols = ["MarkDown1", "MarkDown2", "MarkDown3", "MarkDown4", "MarkDown5"]

dfSales_Features_DataSet_Work[markdown_cols] = dfSales_Features_DataSet_Work[markdown_cols].fillna(0)

# verify
dfSales_Features_DataSet_Work[markdown_cols].isna().sum()

MarkDown1    0
MarkDown2    0
MarkDown3    0
MarkDown4    0
MarkDown5    0
dtype: int64

## Observations - Sales_Features_DataSet

Worked! Thanks Copilot, no missing values in the MarkDown numeric columns

## Observations - Sales_Features_DataSet

Missing values remain the same yet the columns are now without NaN/null values
At this point not concerned, at least the data is clean...

## Strip Spaces From Date Column

In [34]:
#remove spaces from Date column (string datatype)
for objColumn in["Date"]:
    dfSales_Features_DataSet_Work[objColumn] = dfSales_Features_DataSet_Work[objColumn].str.strip()
    
#show first 50 rows to check    
dfSales_Features_DataSet_Work.head(50)    

,Store,Date,Temperature,Fuel_Price,MarkDown1,MarkDown2,MarkDown3,MarkDown4,MarkDown5,CPI,Unemployment,IsHoliday
0,1,05/02/2010,42.31,2.572,0.0,0.0,0.0,0.0,0.0,211.096358,8.106,False
1,1,12/02/2010,38.51,2.548,0.0,0.0,0.0,0.0,0.0,211.242170,8.106,True
2,1,19/02/2010,39.93,2.514,0.0,0.0,0.0,0.0,0.0,211.289143,8.106,False
3,1,26/02/2010,46.63,2.561,0.0,0.0,0.0,0.0,0.0,211.319643,8.106,False
4,1,05/03/2010,46.50,2.625,0.0,0.0,0.0,0.0,0.0,211.350143,8.106,False
5,1,12/03/2010,57.79,2.667,0.0,0.0,0.0,0.0,0.0,211.380643,8.106,False
6,1,19/03/2010,54.58,2.720,0.0,0.0,0.0,0.0,0.0,211.215635,8.106,False
7,1,26/03/2010,51.45,2.732,0.0,0.0,0.0,0.0,0.0,211.018042,8.106,False
8,1,02/04/2010,62.27,2.719,0.0,0.0,0.0,0.0,0.0,210.820450,7.808,False
9,1,09/04/2010,65.86,2.770,0.0,0.0,0.0,0.0,0.0,210.622857,7.808,False


## Observations - Sales_Features_DataSet

No data corruptions

## Validate Numerical Values For Columns Used In Analysis

see if Stores column has values less than 1
see if MarkDown1 column has values less than 1
see if MarkDown2 column has values less than 1
see if MarkDown3 column has values less than 1
see if MarkDown4 column has values less than 1
see if MarkDown5 column has values less than 1


In [35]:
#get count of how many values in Stores column are less than 1
intLessThanOne = (dfSales_Features_DataSet_Work["Store"] < 1).sum()
#print result
intLessThanOne

np.int64(0)

---

## Observations - Sales_Features_DataSet

Values for Store in acceptable limits

## Check All MarkDown Columns For Values Less Than 0

In [36]:
#check MarkDown columns for < 0

for intNum in range(1,6):
    intLessThanOne = (dfSales_Features_DataSet_Work[ f"MarkDown{intNum}" ] < 0).sum()
    #print result
    print( f"MarkDown Column {intNum} Has Values Less Than Zero? {intLessThanOne > 0}")
    
    
#Check findings    
dfSales_Features_DataSet_Work.query( "MarkDown1 <0 or MarkDown2 <0 or MarkDown3 <0 or MarkDown4 <0 or MarkDown5 <0")    

MarkDown Column 1 Has Values Less Than Zero? True
MarkDown Column 2 Has Values Less Than Zero? True
MarkDown Column 3 Has Values Less Than Zero? True
MarkDown Column 4 Has Values Less Than Zero? False
MarkDown Column 5 Has Values Less Than Zero? True


,Store,Date,Temperature,Fuel_Price,MarkDown1,MarkDown2,MarkDown3,MarkDown4,MarkDown5,CPI,Unemployment,IsHoliday
657,4,23/03/2012,59.07,3.759,8806.80,-10.50,5.99,739.14,4396.97,130.896645,4.607,False
860,5,17/08/2012,87.52,3.571,1649.56,-10.98,2.31,1955.75,1205.23,222.627675,5.603,False
873,5,16/11/2012,56.89,3.252,1631.01,-35.74,15.46,326.59,2310.83,224.106624,5.422,False
893,5,05/04/2013,61.88,3.583,9023.29,927.32,170.24,405.30,-185.17,225.682320,5.278,False
1063,6,11/01/2013,48.26,3.243,5510.29,33063.57,-0.86,693.10,3418.77,225.832011,5.372,False
1265,7,31/05/2013,50.70,3.870,2350.78,-7.76,91.59,166.86,540.89,NaN,NaN,False
1587,9,10/08/2012,88.66,3.494,3180.78,-9.94,1.40,2112.03,2314.58,225.717009,5.277,False
1591,9,07/09/2012,87.93,3.730,4837.99,-5.96,22.74,602.80,1377.59,226.210354,5.277,True
1611,9,25/01/2013,49.14,3.227,200.46,348.20,-14.29,59.00,2162.77,228.030618,5.049,False
1615,9,22/02/2013,45.91,3.597,2810.86,1405.04,-179.26,63.12,3593.56,228.205445,5.049,False


## Observations - Sales_Features_DataSet

MarkDown columns 1, 2, 3 and 5 have values less than 0
Being as markdowns can be negative by nature am happy to ignore

## Data Transformations/Feature Engineering

Convert Date column to datetime format

In [37]:
#convert Date column to datetime format
dfSales_Features_DataSet_Work["Date"].astype("datetime64[us]")

#check results
dfSales_Features_DataSet_Work.head(50)

,Store,Date,Temperature,Fuel_Price,MarkDown1,MarkDown2,MarkDown3,MarkDown4,MarkDown5,CPI,Unemployment,IsHoliday
0,1,05/02/2010,42.31,2.572,0.0,0.0,0.0,0.0,0.0,211.096358,8.106,False
1,1,12/02/2010,38.51,2.548,0.0,0.0,0.0,0.0,0.0,211.242170,8.106,True
2,1,19/02/2010,39.93,2.514,0.0,0.0,0.0,0.0,0.0,211.289143,8.106,False
3,1,26/02/2010,46.63,2.561,0.0,0.0,0.0,0.0,0.0,211.319643,8.106,False
4,1,05/03/2010,46.50,2.625,0.0,0.0,0.0,0.0,0.0,211.350143,8.106,False
5,1,12/03/2010,57.79,2.667,0.0,0.0,0.0,0.0,0.0,211.380643,8.106,False
6,1,19/03/2010,54.58,2.720,0.0,0.0,0.0,0.0,0.0,211.215635,8.106,False
7,1,26/03/2010,51.45,2.732,0.0,0.0,0.0,0.0,0.0,211.018042,8.106,False
8,1,02/04/2010,62.27,2.719,0.0,0.0,0.0,0.0,0.0,210.820450,7.808,False
9,1,09/04/2010,65.86,2.770,0.0,0.0,0.0,0.0,0.0,210.622857,7.808,False


## Remove Unnecessary Columns

In [38]:
#going to remove:
#fuel_price, CPI, unemployment
dfSales_Features_DataSet_Work.drop(columns=["Fuel_Price", "CPI", "Unemployment"], inplace=True)

#check results
dfSales_Features_DataSet_Work.info()

<class 'pandas.DataFrame'>
RangeIndex: 8190 entries, 0 to 8189
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Store        8190 non-null   int64  
 1   Date         8190 non-null   str    
 2   Temperature  8190 non-null   float64
 3   MarkDown1    8190 non-null   float64
 4   MarkDown2    8190 non-null   float64
 5   MarkDown3    8190 non-null   float64
 6   MarkDown4    8190 non-null   float64
 7   MarkDown5    8190 non-null   float64
 8   IsHoliday    8190 non-null   bool   
dtypes: bool(1), float64(6), int64(1), str(1)
memory usage: 520.0 KB


## Observations - Sales_Features_DataSet

Columns confirmed removed:

Fuel_Price, CPI, Unemployment

## Save Work DataFrame To CSV File

In [39]:
#create list of columns new and old JUST what is actually needed
lstColumns = [
                #original columns
                "Store","Date","Temperature",
                "MarkDown1","MarkDown2","MarkDown3","MarkDown4","MarkDown5",
                "IsHoliday"
              ]

#create new DataFrame for cleaned data
dfCleaned = dfSales_Features_DataSet_Work[lstColumns].copy()

#save dfSales_DataSet_Work
dfCleaned.to_csv(strCurrentDir + CNST_STR_CLEANFILEPATH)

# Conclusions and Next Steps